### PR TITLE
Add data mapping code action

### PR DESCRIPTION
**Pull request by Ayodhya94**
## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [x] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * Exception for Language Server Code Actions.
+ *
+ * @since 2.0.0
+ */
+public class LSCodeActionProviderException extends Exception {
+
+    public LSCodeActionProviderException(String message) {
+
```
Comment (rasika) [inline language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java:26]: Remove this extra line feed

File: language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        super(message);
+    }
+
+    public LSCodeActionProviderException(String message, Throwable cause) {
+
```
Comment (rasika) [inline language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java:31]: Remove this extra line feed

File: language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        super(message, cause);
+    }
+
+    public LSCodeActionProviderException(Throwable cause) {
+
```
Comment (rasika) [inline language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java:36]: Remove this extra line feed

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    private final boolean enabled;
+    private final String url;
+
+    DataMapperConfig() {
+        this.enabled = false;
+        this.url = "";
+    }
+
+    public boolean isEnabled() {
+
```
Comment (rasika) [inline language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java:16]: Remove this extra line feed

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        return enabled;
+    }
+
+    public String getUrl() {
+
```
Comment (rasika) [inline language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java:21]: Remove this extra line feed, check for other places as well

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+public class DataMapperConfig {
+    private final boolean enabled;
+    private final String url;
+
+    DataMapperConfig() {
+        this.enabled = false;
+        this.url = "";
+    }
+
+    public boolean isEnabled() {
```
Comment (rasika) [inline language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java:15]: Add code-docs for public methods

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+        return enabled;
+    }
+
+    public String getUrl() {
```
Comment (rasika) [inline language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java:20]: Add code-docs for public methods

File: misc/ballerinalang-data-mapper/build-config/checkstyle/checkstyle.xml [hunk 8]
```diff
@@ -0,0 +1,407 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (rasika) [inline misc/ballerinalang-data-mapper/build-config/checkstyle/checkstyle.xml:1]: Any reason to commit this file?

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+     * @param postBody - message payload
+     * @param headers  http request headers map
+     * @return - HttpResponse from end point
+     * @throws IOException If an error occurs while sending the GET request
+     */
+    public static HttpResponse doPost(String endpoint, String postBody, Map<String, String> headers)
+            throws IOException {
+        HttpURLConnection urlConnection = null;
+        try {
+            urlConnection = getURLConnection(endpoint);
```
Comment (rasika) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java:54]: Can we move this part into try-with-resources approach? So we can remove https://github.com/ballerina-platform/ballerina-lang/pull/25022/files#diff-c9d62fe95c71192e4a81b263dc056364R66 part

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        this.responseCode = responseCode;
+    }
+
+    public HttpResponse(String data, int responseCode, Map<String, String> headers) {
+        this.data = data;
+        this.responseCode = responseCode;
+        this.headers = headers;
+    }
+
+    public String getData() {
```
Comment (rasika) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java:36]: Add code-docs for public methods(even you copied it from somewhere else)

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        return data;
+    }
+
+    public int getResponseCode() {
```
Comment (rasika) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java:40]: Add code-docs for public methods

File: misc/ballerinalang-data-mapper/src/test/java/org/ballerinalang/datamapper/codeaction/CodeActionTest.java [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+    @BeforeClass
+    private void init() {
+        this.serviceEndpoint = TestUtil.initializeLanguageSever();
+        String startConfigPath = "codeaction" + File.separator + "config" + File.separator + "startConfig.json";
+        JsonObject configs = FileUtils.fileContentAsObject(startConfigPath);
+        TestUtil.setWorkspaceConfig(serviceEndpoint, configs);
+    }
+
+    @Test(dataProvider = "codeAction-data-mapper-data-provider")
+    public void testDataMapperCodeAction(String config, String source) throws Exception {
```
Comment (rasika) [inline misc/ballerinalang-data-mapper/src/test/java/org/ballerinalang/datamapper/codeaction/CodeActionTest.java:78]: this test method is bit long, commenting would helpful for  readability

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
+package org.ballerinalang.langserver.compiler.config;
```
Comment (nadeeshaan) [inline language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java:1]: Missing license agreement

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/LSClientConfig.java [hunk 14]
```diff
@@ -105,4 +107,9 @@ public boolean isTraceLogEnabled() {
     public GoToDefinitionConfig getGoToDefinition() {
         return goToDefinition;
     }
+
+    public DataMapperConfig getDataMapper() {
+
```
Comment (nadeeshaan) [inline language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/LSClientConfig.java:112]: format needed

File: misc/ballerinalang-data-mapper/src/test/resources/testng.xml [hunk 15]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (miyurud) [inline misc/ballerinalang-data-mapper/src/test/resources/testng.xml:3]: Change the copyright statement here.

File: language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java (hunk 1 above)
Comment (Ayodhya94) [reply language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java:26]: Done

File: language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java (hunk 2 above)
Comment (Ayodhya94) [reply language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java:31]: Done

File: language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java (hunk 3 above)
Comment (Ayodhya94) [reply language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/LSCodeActionProviderException.java:36]: Done

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java (hunk 4 above)
Comment (Ayodhya94) [reply language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java:16]: Done

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java (hunk 5 above)
Comment (Ayodhya94) [reply language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java:21]: Done

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java (hunk 6 above)
Comment (Ayodhya94) [reply language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java:15]: Done

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java (hunk 7 above)
Comment (Ayodhya94) [reply language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java:20]: Added code-docs

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java (hunk 13 above)
Comment (Ayodhya94) [reply language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java:1]: Added license agreement

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/LSClientConfig.java (hunk 14 above)
Comment (Ayodhya94) [reply language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/LSClientConfig.java:112]: Formatted

File: misc/ballerinalang-data-mapper/src/test/resources/testng.xml (hunk 15 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/test/resources/testng.xml:3]: Changed the statement.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java (hunk 11 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java:37]: Added code-docs

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java (hunk 10 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java:33]: Done

File: misc/ballerinalang-data-mapper/src/test/java/org/ballerinalang/datamapper/codeaction/CodeActionTest.java (hunk 12 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/test/java/org/ballerinalang/datamapper/codeaction/CodeActionTest.java:78]: Done

File: misc/ballerinalang-data-mapper/build.gradle [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+    implementation project(':language-server:language-server-commons')
+    implementation project(':language-server:language-server-compiler')
+    implementation project(':language-server:language-server-core')
+    implementation project(':language-server:language-server-cli')
+    implementation project(':openapi-ballerina:openapi-validator')
+    implementation project(':openapi-ballerina:openapi-to-ballerina-generator')
+    implementation project(':openapi-ballerina:ballerina-to-openapi-generator')
+
+    // Standard libs
+    implementation project(':ballerina-auth')
```
Comment (nadeeshaan) [inline misc/ballerinalang-data-mapper/build.gradle:33]: We might not need all of these modules, let's remove them as soon as we merge the PR

Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/build.gradle:33]: Okay.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeAction.java [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+ *
+ * This software is the property of WSO2 Inc. and its suppliers, if any.
+ * Dissemination of any information or reproduction of any material contained
+ * herein is strictly forbidden, unless permitted by WSO2 in accordance with
+ * the WSO2 Commercial License available at http://wso2.com/licenses.
+ * For specific language governing the permissions and limitations under
+ * this license, please see the license as well as any agreement you’ve
+ * entered into with WSO2 governing the purchase of this software and any
+ */
+package org.ballerinalang.datamapper;
```
Comment (anupama-pathirage) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeAction.java:12]: It seems the license header is not properly added. The last part is missing and also new lines are missing. Refer [1] for example.

[1] https://github.com/ballerina-platform/ballerina-lang/blob/master/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/preloading/LSClientConfig.java

File: language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java (hunk 13 above)
Comment (anupama-pathirage) [reply language-server/modules/langserver-compiler/src/main/java/org/ballerinalang/langserver/compiler/config/DataMapperConfig.java:12]: It seems the license header is not properly added. The last part is missing and also new lines are missing. Refer [1] for example.

[1] https://github.com/ballerina-platform/ballerina-lang/blob/master/tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/preloading/LSClientConfig.java

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.com). All Rights Reserved.
+ *
+ * This software is the property of WSO2 Inc. and its suppliers, if any.
+ * Dissemination of any information or reproduction of any material contained
+ * herein is strictly forbidden, unless permitted by WSO2 in accordance with
+ * the WSO2 Commercial License available at http://wso2.com/licenses.
+ * For specific language governing the permissions and limitations under
+ * this license, please see the license as well as any agreement you’ve
+ * entered into with WSO2 governing the purchase of this software and any
+ */
```
Comment (anupama-pathirage) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:11]: Same as above. Check all the places.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+        JsonObject rightRecordJSON = new JsonObject();
+        JsonObject leftRecordJSON = new JsonObject();
+
+        // Schema 1
+        BType variableTypeMappingFrom = symbolAtCursor.type;
+        if (variableTypeMappingFrom instanceof BRecordType) {
+            List<BField> rightSchemaFields = new ArrayList<>(((BRecordType) variableTypeMappingFrom).fields.values());
+            JsonObject rightSchema = (JsonObject) recordToJSON(rightSchemaFields);
+
+            rightRecordJSON.addProperty("schema", foundTypeRight);
```
Comment (anupama-pathirage) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:143]: Can we define set of constants for these fields as we are using them in multiple places. Then we can find the usages easily and change easily in future if requrired.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+        conn.setConnectTimeout(15000);
+        conn.setDoInput(true);
+        conn.setUseCaches(true);
+        conn.setAllowUserInteraction(false);
+        return conn;
+    }
+
+    private static Map<String, String> readHeaders(URLConnection urlConnection) {
+        Iterator<String> itr = urlConnection.getHeaderFields().keySet().iterator();
+        Map<String, String> headers = new HashMap();
```
Comment (anupama-pathirage) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java:86]: Should be `= new HashMap<>();`

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+                urlConnection.disconnect();
+            }
+        }
+    }
+
+    private static HttpURLConnection getURLConnection(String requestUrl) throws IOException {
+        return getURLConnection(requestUrl, DEFAULT_READ_TIMEOUT);
+    }
+
+    private static HttpURLConnection getURLConnection(String requestUrl, int readTimeout) throws IOException {
```
Comment (anupama-pathirage) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java:72]: Why `readTimeout` is passed as a parameter as it is constant. Also for connect time out also, we have used a constant value(15000) but it is not passed as a parameter or defined as a constant. Can we follow the same convention for all the properties?

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+            conn.addRequestProperty(e.getKey(), e.getValue());
+        }
+        conn.setRequestMethod(method);
+    }
+
+    private static HttpResponse buildResponse(HttpURLConnection conn) throws IOException {
+        return buildResponse(conn, defaultResponseBuilder, false);
+    }
+
+    private static HttpResponse buildResponse(HttpURLConnection conn,
```
Comment (anupama-pathirage) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java:108]: We have only one usage of `buildResponse` function. So why do we need pass `throwError` as a parameter which is always `false`.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+     * This method set data and response code of a HTTP request response.
+     * @param data {@link String}
+     * @param responseCode {@link int}
+     */
+    public HttpResponse(String data, int responseCode) {
+        this.data = data;
+        this.responseCode = responseCode;
+    }
+
+    HttpResponse(String data, int responseCode, Map<String, String> headers) {
```
Comment (anupama-pathirage) [inline misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java:33]: Input parameter `headers` are unused.

File: misc/ballerinalang-data-mapper/src/test/resources/codeaction/source/dataMapper1.bal [hunk 24]
```diff
@@ -0,0 +1,28 @@
+type Grades record {|
```
Comment (anupama-pathirage) [inline misc/ballerinalang-data-mapper/src/test/resources/codeaction/source/dataMapper1.bal:1]: Please add license header for all .bal files as well. Check all the places.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeAction.java (hunk 17 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeAction.java:12]: Changed the license header as suggested

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java (hunk 18 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:11]: Changed the license headers as suggested

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java (hunk 20 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java:86]: Changed

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java (hunk 21 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java:72]: Made the changes to follow the same convention.

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java (hunk 22 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpClientRequest.java:108]: Removed unwanted parameter

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java (hunk 23 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/utils/HttpResponse.java:33]: Removed unwanted parameters

File: misc/ballerinalang-data-mapper/src/test/resources/codeaction/source/dataMapper1.bal (hunk 24 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/test/resources/codeaction/source/dataMapper1.bal:1]: Added license headers

Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/test/resources/codeaction/source/dataMapper1.bal:1]: Added to all the .bal files

File: misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java (hunk 19 above)
Comment (Ayodhya94) [reply misc/ballerinalang-data-mapper/src/main/java/org/ballerinalang/datamapper/AIDataMapperCodeActionUtil.java:143]: Defined them as constants

**Pull request merged**